patches/minecraft/net/minecraft/client/renderer/ItemBlockRenderTypes.java.patch

```diff
@@ -54,50 +54,78 @@
 +
 +   // FORGE START
 +
-+   private static final Map<net.minecraftforge.registries.IRegistryDelegate<Block>, java.util.function.Predicate<RenderType>> blockRenderChecks = Maps.newHashMap();
-+   private static final Map<net.minecraftforge.registries.IRegistryDelegate<Fluid>, java.util.function.Predicate<RenderType>> fluidRenderChecks = Maps.newHashMap();
++   private static final Map<net.minecraftforge.registries.IRegistryDelegate<Block>, java.util.function.Predicate<RenderType>> blockRenderChecks = createRenderCheckMap();
++   private static final Map<net.minecraftforge.registries.IRegistryDelegate<Fluid>, java.util.function.Predicate<RenderType>> fluidRenderChecks = createRenderCheckMap();
++   private static final java.util.concurrent.locks.StampedLock RENDER_CHECK_LOCK = new java.util.concurrent.locks.StampedLock();
 +   static {
 +      f_109275_.forEach(ItemBlockRenderTypes::setRenderLayer);
 +      f_109276_.forEach(ItemBlockRenderTypes::setRenderLayer);
 +   }
 +
++   private static <T> Map<net.minecraftforge.registries.IRegistryDelegate<T>, java.util.function.Predicate<RenderType>> createRenderCheckMap() {
++      java.util.function.Predicate<RenderType> solidPredicate = type -> type == RenderType.m_110451_();
++      return Util.m_137469_(new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(), map -> map.defaultReturnValue(solidPredicate));
++   }
++
 +   public static boolean canRenderInLayer(BlockState state, RenderType type) {
 +      Block block = state.m_60734_();
 +      if (block instanceof LeavesBlock) {
 +         return f_109277_ ? type == RenderType.m_110457_() : type == RenderType.m_110451_();
 +      } else {
-+         java.util.function.Predicate<RenderType> rendertype;
-+         synchronized (ItemBlockRenderTypes.class) {
-+            rendertype = blockRenderChecks.get(block.delegate);
-+         }
-+         return rendertype != null ? rendertype.test(type) : type == RenderType.m_110451_();
++         return canRenderInLayer(type, blockRenderChecks, block.delegate);
 +      }
 +   }
 +
 +   public static boolean canRenderInLayer(FluidState fluid, RenderType type) {
-+      java.util.function.Predicate<RenderType> rendertype;
-+      synchronized (ItemBlockRenderTypes.class) {
-+         rendertype = fluidRenderChecks.get(fluid.m_76152_().delegate);
++      return canRenderInLayer(type, fluidRenderChecks, fluid.m_76152_().delegate);
++   }
++
++   private static <T> boolean canRenderInLayer(RenderType type, Map<T, java.util.function.Predicate<RenderType>> map, T key) {
++      java.util.concurrent.locks.StampedLock lock = RENDER_CHECK_LOCK;
++
++      // attempt optimistic read to the render type predicate without blocking
++      long stamp = lock.tryOptimisticRead();
++      if (stamp != 0) {
++         java.util.function.Predicate<RenderType> predicate = map.get(key);
```

Comment:
Does this not risk crashing if the map is being modified concurrently?

Reply:
A get query should not throw an exception with concurrent access (at least, with `Object2ObjectOpenHashMap`)- but it might return invalid data, though this is protected by the stamp validation.
From previous testing, though it was a while ago, acquiring a read lock instead of performing an optimistic read was significantly more expensive, mainly due to having to acquire and mutate the lock state in such hot code. Using a copy-on-write map might be a much more elegant solution, though! Would this be preferred? 🙂

Reply:
> A get query should not throw an exception with concurrent access

Afaik, all bets are off when it comes to concurrent access with fastutil maps. :P

The fancy solution would be to have a `volatile boolean modified` that is set after each modification, and have read synchronize then copy the map if the flag is set. Possibly a bit overkill. :P

Reply:
It's definitely not ideal to rely on fastutil internals to not throw any exceptions. 😄 
Might it make sense to re-assign the map right away after a render type is set, just to avoid any overhead in the lookup logic, even if it means setting into the map is a bit more expensive?